src/ring_theory/derivation.lean

```diff
@@ -0,0 +1,223 @@
+/-
+Copyright © 2020 Nicolò Cavalleri. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Nicolò Cavalleri.
+-/
+
+import algebra.lie_algebra
+import tactic
+
+/-!
+# Derivations
+
+This file defines derivation. A derivation from the `R`-agebra `A` to the `A`-module `M` is an `R`
+linear map that satisfy the Leibniz rule.
+
+## Notation
+
+The notation `⁅D1, D2⁆` is used for the commutator of two derivations.
+
+TODO: this file is just a stub to go on with some PRs in the geometry section. It only
+implements the definition of derivations in commutative algebra. This will soon change: as soon
+as bimodules will be there in mathlib I will change this file to take into account the
+non-commutative case. Any development on the theory of derivations is discouraged until the
+definitive definition of derivation will be implemented.
+
+Also note that this file implements bundled derivations only by now. To be better integrated with
+the algebra section of mathlib it would probably be better to define a notion of `is_derivation` and
+link it to bundled derivations.
```

Comment:
I believe that the consensus among experts (i.e. people who know more about it Lean than me) is that maps (`ring_hom`, `linear_map`, ...) and subobjects (`subgroup`, `submodule`) work better when they are bundled. Therefore, I would not expect that it would help to define `is_derivation`. (The drawback is that in the end we will have to repeat a lot of the API for linear maps to make it work for derivations.)

Reply:
Well I added this after Oliver's comment. I personally also prefer to talk about substructures by talking about the inclusion morphisms and never mention things as subsets or in this case subtypes, but if Oliver prefers the other approach I am totally fine with it: I see that one might want to talk explicitly of subalgebras instead inclusion morphisms, it is also a matter of taste! In any case then I'll remove this comment then, for sure I will not implement `is_derivation` myself, so it will his / other people choice.